core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala

```diff
@@ -0,0 +1,727 @@
+
+          // the entry has timed out; if the active ack is still around, remove its entry also
+          // and complete the promise with a failure if necessary
+          activationPromises
+            .remove(aid)
+            .foreach(_.tryFailure(new Throwable("no completion or active ack received yet")))
+        }
+
+      // Active acks that are received here are strictly from user actions - health actions are not part of
+      // the load balancer's activation map. Inform the invoker pool supervisor of the user action completion.
+      case None if !forced =>
+        // Received a completion ack that has already been taken out of the state because of a timeout (forced ack).
+        // The result is ignored because a timeout has already been reported to the invokerPool per the force.
+        // Logging this condition as a warning because the invoker processed the activation and sent a completion
+        // message - but not in time.
+        logging.warn(this, s"received completion ack for '$aid' which has no entry, system error=$isSystemError")(tid)
+      case None =>
+        // The entry has already been removed by a completion ack. This part of the code is reached by the timeout and can
+        // happen if completion ack and timeout happen roughly at the same time (the timeout was triggered before the completion
+        // ack canceled the timer). As the completion ack is already processed we don't have to do anything here.
+        logging.debug(this, s"forced completion ack for '$aid' which has no entry")(tid)
+    }
+  }
+
+  private val queueKey = QueueKeys.queuePrefix
+  private val schedulerKey = SchedulerKeys.prefix
+  private val throttlingKey = ThrottlingKeys.prefix
+  private val watchedKeys = Set(queueKey, schedulerKey, throttlingKey)
+
+  private val watcher = actorSystem.actorOf(Props(new Actor {
+    watchedKeys.foreach { key =>
+      watcherService ! WatchEndpoint(key, "", true, watcherName, Set(PutEvent, DeleteEvent))
+    }
+
+    override def receive: Receive = {
+      case WatchEndpointRemoved(watchKey, key, value, true) =>
+        watchKey match {
+          case `queueKey` =>
+            if (key.contains("leader")) {
+              queueEndpoints.remove(key)
+              activationSlots.values
+                .find(entry =>
+                  //the leader key's value is queue/invocationNamespace/ns/pkg/act/leader
+                  QueueKeys
+                    .queue(entry.invocationNamespace, entry.fullyQualifiedEntityName.copy(version = None), true) == key)
+                .foreach {
+                  entry =>
+                    implicit val transid = entry.transactionId
+                    logging.warn(
+                      this,
+                      s"The $key is deleted from ETCD, but there are still unhandled activations for this action, try to create a new queue")
```

Comment:
Trying to remember the architecture here, but how do unhandled activations still get retried after creating the new queue. Is it because this load balancer puts them on kafka and they'd just be sitting there waiting to be consumed by the scheduler with the recreated queue still? If the etcd queue gets deleted with an activation in it, then that's just lost at that point right since it's already been read off kafka?

Reply:
When there is an activation in a queue, the queue is not deleted.
Unhandled activations here come from the possibility that while the queue is deleted because of timeout(no more activation), still there can be new activations incoming on the controller side.

Regarding the role of a controller, yes, now it is rather passive.
It just selects the right scheduler and send a queue creation request and put activations to one of the scheduler's Kafka topic.
Still, activations can be delivered to a scheduler without a queue for the given activations, then the scheduler looks for(with retries) the target scheduler that has a queue and forwards activations to the scheduler.
(I hope we can remove Kafka from the critical path entirely in the future. Then we are probably able to consolidate these two calls into one.)

So basically the situation here is supposed to rarely happen.
Once a queue is created, the activation will be forwarded to the proper scheduler.